tests/test_controllers.py

```diff
@@ -22,10 +22,11 @@ def test_optimize_points_in_extraction_model(mocker):
     em.add_point('0', 0, 75, 85)
 
     mocker.patch('bmlab.session.Session.extraction_model', return_value=em)
+    # mocker.patch('bmlab.session.Session.get_instance().current_repetition().calibration.get_image', return_value=img)
```

Comment:
@matthias-baer Since `ec.optimize_points` now gets the image from the Session, a session would be required for the test to run. Is it instead possible to mock `bmlab.session.Session.get_instance().current_repetition().calibration.get_image()` instead to return `img`? My naive (commented) approach here did not work.

Reply:
Well, I guess you would have to mock the three objects (Session, Repetition, Calibration) separately.  The reason that this gets so cumbersome here is that we are violating Demeter's law. It would probably be neater to have a method in Sessin along the lines `session.get_calibration_image(repetition_key=None)` which returns the calibration image directly (None=>current repetition), so that the user of session does not have to know about the internals of Session/Repetition/Calibration classes. What do you think?

Reply:
> 
> 
> Well, I guess you would have to mock the three objects (Session, Repetition, Calibration) separately. The reason that this gets so cumbersome here is that we are violating Demeter's law. It would probably be neater to have a method in Sessin along the lines `session.get_calibration_image(repetition_key=None)` which returns the calibration image directly (None=>current repetition), so that the user of session does not have to know about the internals of Session/Repetition/Calibration classes. What do you think?

Yes, we still violate this principle in a couple of places. That's why we are here 🙈 🙂

I also thought about having methods in the session which provide the necessary stuff. I even think we don't have to specify the repetition, as one only ever requires data from the current repetition (but we need to specify the `calibration_key`).
I will implement a method `session.get_calibration_image(calib_key)` which returns calibration images for the current repetition. Is that ok for you?

Later, we will also require methods for the payload images, and we also need meta-data from the images (acquisition date, exposure duration, maybe more).

Reply:
These methods can then also already take care of the correct orientation.